src/asymmetric_crypto/curve_25519/ed25519/ed_dsa/key_pair.rs

```diff
@@ -100,3 +118,52 @@ impl Ed25519Keypair {
         })
     }
 }
+
+/// Algorithm [`ObjectIdentifier`] for the Ed25519 digital signature algorithm
+/// (`id-Ed25519`).
+///
+/// <http://oid-info.com/get/1.3.101.112>
+pub const ED25519_ALGORITHM_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.101.112");
+
+/// Ed25519 Algorithm Identifier.
+pub const ED25519_ALGORITHM_ID: pkcs8::AlgorithmIdentifierRef<'static> =
+    pkcs8::AlgorithmIdentifierRef {
+        oid: ED25519_ALGORITHM_OID,
+        parameters: None,
+    };
+
+impl EncodePublicKey for Ed25519Keypair {
+    fn to_public_key_der(&self) -> spki::Result<Document> {
+        pkcs8::SubjectPublicKeyInfoRef {
+            algorithm: ED25519_ALGORITHM_ID,
+            subject_public_key: BitStringRef::new(0, self.public_key.as_bytes())?,
+        }
+        .try_into()
+    }
+}
+impl EncodePrivateKey for Ed25519Keypair {
+    fn to_pkcs8_der(&self) -> pkcs8::Result<pkcs8::SecretDocument> {
+        // Serialize private key as nested OCTET STRING
+        let mut private_key = [0u8; 2 + (Self::LENGTH / 2)];
+        private_key[0] = 0x04;
+        private_key[1] = 0x20;
+        private_key[2..].copy_from_slice(self.private_key.as_bytes());
```

Comment:
are we confident `Self::LENGTH / 2` equals to `self.private_key.as_bytes().len()`? Otherwise it may panic

Reply:
This works because both the public and private keys are the same length. But this is obfuscating.
I would use the length of the private key instead.

Reply:
it has been fixed using the constant CURVE_25519_PRIVATE_KEY_LENGTH